Example of system decomposition (#36)
elisa-tech/wg-osep: Contributions/Linux_Kernel_Safety_First_Principles.md

## **Structure of the document**
The next section illustrates the philosohpy behind the approach used here to
deal with safety requirements vs a system using Linux.
The further two sections about "First Principles" address two aspects of safety:
-  integrity: reliably detecting interference to components with safety requirements
-  availability: protracted integrity of the safey-relevant functionality
The last section lists considerations that can help with raching the safety goals.

## **Philosophy of the Document**
*  A system subject to safety goals might employ the Linux Kernel.
*  Safety goals are specific to the overall system and the related use cases.
*  Similarly, one must expect that the solution will be tailored to both system and goals.
*  However, it is possible to describe the Linux Kernel from a safety point of view.
*  Not all the issues listed will be necessarily relevant to each system/application.
*  A Safety Analysis should, however, clearly state which issues are relevant, which are not, and why.

## **First Principles - Integrity**
This section provides touchstone concepts about interference of QM components with
safety-qualified ones.

> 1. The vanilla Linux Kernel, is not sufficient to support, alone, safey goals.

Reviewer: IMO this statement is misleading as it can be read as a negative conclusion. IMO it should be rephrased as "To support safety goals, The vanilla Linux Kernel must be accompanied by appropriate evidences/argumentations"

Author: I think I'm stating the facts, while you are making the assumption that such evidence /argumentation exists
Burden of proof is on you, then. :-) 
Till the evidence is provided and verified, this is the situation.

> 2. No internal kernel barriers/protections against interference to variable data, including safety-relevant data.

Reviewer: What does this mean? variables stored in stack for example can be protected by CONFIG_STACKPROTECTOR to some degree. IMO we should say "the monolithic nature of the Kernel poses challenges in protecting safety-relevant data against interference"

Author: A barrier is a barrier. A barrier is something that does not let the interference pass.
STACKPROTECTOR is an after-the-fact mechanism.
And notice that stackprotector doesn't protect anything, in practice.
"Stackdetectector" would be a better name.

3. Any component within the kernel context can generate (cascaded) interference.

> 4. Internal interference can cascade also through components that have been qualified for safety at unit-level.

Reviewer: I would remove "at unit-level" as the concept of unit, usually, applies to newly developed SW, not when qualifying pre-existing one

Author: unit is meant to represent the granularity chosen for the qualification, whatver that might be.
I'm trying to say SeooC without using the terminology, but that is the intended meaning.
As long as he meaning is preserved, I'm open to re-formulating it.

Reviewer 2: ```suggestion
4. Internal interference can also cascade through components that have only been safety-qualified at a component level.
```

Author: @reiterative I'm almost ok with this, but I do not like that "only" because it seems to imply that something can be done on components to prevent the cascading of interference.
In theory yes, you could add redundance and what not, but in practice I don't think anyone will be able to make components sufficiently hardened.
If that is what you are after, then let's spell it out. I want to paint an accurate picture of the problem, without closing avenues that are open, but I do not want to present a rocky mountain path as if it was a boulevard either :-D

5. The Linux Kernel is able to interfere with both itself and any part of user-space processes (linear map).

> 6. No generic solution for very complex systematic corruption of any writable memory.

Reviewer: Here what do you mean by "generic solution"? Do you mean "currently there is no design solution to protect writable safety relevant memory"?

Author: "Currently" is implied, but, again, as long as something is proven to work, it's an absolute statement.
It can be rephrased, if it helps with readability, without altering the meaning.
My point is that the "Very complex systematic corruption" cannot be detected through solutions that are applicable to - let's say - the memory map of a safety-relevant process, that might comprise tens, hundreds of MB of memory. You can certainly try to use ad-hoc verification, e.g. performing functional testing on that process, but it needs to be tailored to each individual process. No generic solution. Only a generic methodology, to be taylored.

Reviewer 2: I think we can make a stronger (but narrower) statement here:
```suggestion
6. Complex systematic corruption of writable memory cannot be prevented by a generic, kernel-level solution
```

Author: @reiterative: "neither prevented nor detected" is the right formualtion, I think

Reviewer: @igor-stoppa  I think your statement is not correct. For example CONFIG_STACKPROTECTOR_STRONG or KASAN are generic Kernel solution that provide detections of writable memory. So I would stick with Pauls' rewording

Author: @paolonig They are partial solutions. Placing canaries on the stack doesn't catch all the possible corruptions.
KASAN alters the system configuration, and anyways is not deployed.
Feel free to provide evidence countering my arguments, they are intentionally made in a very clan cut way, to allow easy confutation by evidence, if any can be found.

Reviewer: Yes they are partial generic kernel level solutions, I agree. However an integrator or a SW vendor may use such partial solution in combination with other measures to support a non-interference claim.
So IMO the fact is the monolithic nature of the Kernel that does not allow, today, interference prevention (like the one you can have by SMMU for a user space process) and we can all agree on this.
What I cannot agree is about principles that also imply or deny specific ways to support a certain claim; since the specific ways depend on the specification of the integration context and of other projects aspects (depending on the engineering workforce I may decide to use some verification strategies rather than others....)

> 7. Stress testing alone is not sufficient for supporting safety claims.

Reviewer: This is conclusion that IMO is not appropriate. An OEM could stress test the OS in a car for as many hours as needed to claim proven in use for example. IMO it can be reprhased as "It is usually difficult to support a safety claim exclusively by stress testing"

Author: No. Stress testing is certainly beneficial, but alone it is not sufficient.
Or else you are saying that QM practices can turn something into ASIL.
I'm sure you don't mean that.

> 8. Safety claims on the Linux Kernel require both positive testing and focused negative testing.

Reviewer: Also here the statement is too absolutist. I would rephrase as "For requirement based testing usually both positive testing as well as focused negative testing are needed"

Author: It is trying exactly to draw a line between claims that are sound and claims that are not.
Of course one can make a leap of faith, cross the fingers and hope.
But without focused negative testing it is impossible to assess that safety claims on the kernel (alone) are valid.
Of course one can choose other mechanisms, e.g. redundancy, but that is a completely different system.
Once more, I'm happy to be proven wrong, but without verifiable counter evidence, the claim stands.

Reviewer 2: I concur that both positive and negative testing are required. I would phrase it like this:
```suggestion
8. Where safety claims involving the Linux Kernel are supported by testing, both positive, requirements-based testing and focused, analysis-led negative testing are required.
```

9. Those security features which are based on randomisation decrease repeatability of testing (e.g. structure layout randomisation).

> 10. Safety claims must be supported by components with the same-or-better safety level (e.g. a safety-qualified watchdog).

Reviewer: incomplete sentence IMO. I would say: Safety claims must be supported by components with the same-or-better safety level than the ones associated with them (e.g. a safety-qualified watchdog)

Author: Isn't it obvious that there is an association?

Reviewer 2: ```suggestion
10. If safety claims relating to one component are based on support by another, then the supporting component must have the same or better level of safety integrity (e.g. a safety-qualified watchdog).
```

> 11. Non safety qualified processes can interfere with safety qualified ones, indirectly, through the kernel (e.g. triggering memory management bugs).

Reviewer: IMO this sentence expose the final (bad) effect without focusing on the topic. I would say: The Kernel plays a key role for supporting the claim of coexistence of processes of different criticality (e.g. a memory management bug could trigger interferences between such processes).

Author: The topic is the interference (AKA final bad effect).
The kernel doesn't really support anything, being QM at best.
Again, waiting for objective counter evidence.

Reviewer 2: I think a stronger statement about the nature of the risk is clearer here:
```suggestion
11. Any other process could potentially interfere with a safety-qualified process, indirectly, through the kernel (e.g. a memory management bug could trigger interference between such processes).
```

> 12. Using cgroups/containers/SELinux removes only certain simple types of user-space-induced interference.

Reviewer: I would remove "simple" as it is too much debatable IMO

Author: simple -> direct ?

> 13. In a mixed-criticality scenario, non safety-qualified code represents a safety liability that grows with the frequency of execution (cgroups/containers, LSM/SELinux, etc.).

Reviewer: Are you saying that the more often NSR code is executed the higher the chance of interference is? If so why are you mentioning cgroups/containers ?

Author: NSR = Non Safety Relevant? 
If that is the case, then no, I really mean qualified, as out of context.
cgroups are not qualified - and qualifying them would mean qualifying the entire kernel, more or less
same goes for LSM/SELinux

The fact that one might want to use them to prevent a certain type of interference dpoesn't automagically qualify them :-D

Reviewer: @reiterative  I would rephrase as <<In a mixed-criticality scenario, all unqualified code represents a potential risk of interference, which grows with frequency of execution; therefore any code supporting the mixed criticality safety claim (e.g. cgroups/containers, LSM/SELinux, etc.) needs to be qualified according to the highest safety integrity level of the supported safety processes>>

The reason being that the way it is currently written sounds like "cgroups/containers, LSM/SELinux" are bad actors; instead I think we should focus on what needs to be done if we want to leverage such features...

Author: But they *ARE* bad actors, since they are not ASIL
Again, evidence supporting the opposite is welcomed.
Till then, I think I stated the existing situation.

Reviewer: Then any piece of the Kernel is a bad actor unless you find all the evidence that you need to build a safety case.
IMO the goal of this doc should be to help focusing on the challenges that must be solved for using the Kernel in supporting certain safety claim. So if you need to claim mixed criticality between user space processes you need to focus on the qualification of MM, Scheduler, namespaces, cgroups, SELinux etc. 
Instead if you need to claim FFI between Kernel threads (that also includes interference of a QM driver towards a user space process), you still need to focus on the qualification of MM and scheduler plus you need to model the interference from a QM driver toward an ASIL kernel component and toward a user mode thread.

Author: Indeed: there is no reason to expect any safety level from any prt of the kernel, unless proven differently.

The goal of this document is simply to provide *facts* that one can start from.
Speculations can certainly be made available, but not here.

14. HW enhancements are not a catch-all solution (e.g. ECC Memory doesn't prevent interference from non-safety-qualified SW).

> 15. Detection of interference depends more on the recipient than the origin of the interference: too many sources, too difficult to model reliably.

Reviewer: I don't think it is possible to agree in general with such a statement. E.g. KASAN would detect interference at the origin, not at the recipient...

Reviewer 2: ```suggestion
15. When there are many possible sources of interference in a system, we can only reliably model and detect it in the receiving context.
```

Reviewer: @igor-stoppa  maybe a specific integrator decides that the associated performance drop is OK...you dunno 100% for sure, therefore @reiterative  I would rephrase as 

<<When there are many possible sources of interference in a system, it is usually easier to model and reliably detect it in the receiving context>>

Author: and, again, it's not about performance, it's about completeness: you need to present not only the sources covered, but also the ones NOT covered, to claim a certain level of coverage.
And it's the latter part that is practically impossible.
Again, if it is possible, feel free to prove me wrong.

> 16. Difficult to say when all necessary resources are allocated and if they will be retained (e.g. process memory pages).

Reviewer: If the concern here is time interference I think the problem does not limit to resource retention only. I would say that, <<given the current complexity of HW SoCs and of the programs running on top, it is hard to claim that a safety workload is able to meet the allocated FDTI with a certain ASIL rigor without an external monitor>>

Author: The problem is not with the HW, in fact here are systems which are qualified.
Put Linux on the same HW and this problem arises.
It's really an OS problem.

Reviewer 2: ```suggestion
16. The precise timing of allocation (e.g. during system or process startup) and retention of dynamically managed shared resources (e.g. process memory pages) is difficult to predict in a Linux-based system
```

## **First Principles - Availability**
This section provides touchstone concepts about (lack of) system availability, in presence of non-safety-qualified components.

1. Detecting interference alone doesn't help with controlling/managing availability.

> 2. The nature of the very complex systematic interference makes estimating probability of a failure unrealistic.

Reviewer: I disagree on this point. For instance even if an external safety monitor is required to meet the FDTI claim, it is still possible to tune the system and stress test it to claim that the unavailability risk due to temporal interference is acceptable

Author: Feel free to estimate in a verifiable way said probability, then.
We can review the soundness of the estimation.

Reviewer: You can do that only if you define what is a good-enough RT performance. BTW you can start from [this paper](https://bristot.me/files/research/papers/ecrts2020/deOliveira2020Demystifying.pdf) and the toolset mentioned within it as a good starting point.

Author: I'm not talking about RT performance, at least not in the time scale commonly intendend. 
I'm talking about minutes (or more) overheads to basic operations.

Reviewer: Can you make an example of a basic operation and also explaining what is preventing it from running?

Author: Imagine a set of processes implementing machine vision using tens of GB or ram.
Try to monitor that in realtime for corruption caused by the kernel ...

> 3. Stress testing as a means of safety qualification might be realistic at most for very simple cases.

Reviewer: But here the topic is availability, not safety qualification....

Reviewer: Most of the use cases I worked on did not require the OS to be safe operational. I.e. panicing the OS would lead the system in an acceptable safe state (with a very upset driver, hence the need for availability, that however is not related to safety)

Author: Ok, then just keep the system off :-)

Reviewer: Jokes apart I think that in order to agree to such principles we need first to define the integration context.
As I mentioned all the customers I worked with usually accept to have an availability test campaign to be separate from safety (since safety is guaranteed by the external monitor in case of missing the FDTI deadline)

Author: These principles are meant to not be democratically decided by what anyone customer's might think.
I understand "the customer is always right" approach, but that doesn't belong to this document.

Anyone is free to take the risks they prefer, small or large.

This document is not about that. It's about separating assessments that can be done with very high confidence, if not certainty, from others that are based on subjective analysis.

For example, can any of your customers prove that they have determined the coverage of that stress testing?

Performing the same or similar operation billions of times doesn't say anything about other scenario that might arise.
It can be done for sure in simple scenarios, but in more complex ones it cannot.
Again, I'm tryin to make it easy to prove me wrong, by making fairly clean cut statements.

Just add evidence ...

4. Qualitatively, the likelihood of a failure grows with invocativation of non-safety-qualified components. For example:
   1. invocation of device drivers, especially if not related to safety operations
   2. syscalls
   3. memory allocation/release operations
   4. I/Os, e.g. storage, network
   5. evaluation of Linux Security Module hooks
   6. evaluation of cgroups tests
   7. presence of non-safety-relevant triggering many of the above points

> 

Reviewer: Above points are correct, and this is why a QM baseline is always required to support availability, but the same baseline applies to NSR components....so what is the goal here? Wouldn't it be easier to say "a QMS baseline is required to meet the products' availability requirements" ?

Author: The point is that a QM component is by definition less safe than an ASIL one.
Even in presence of perfect detection of interference QM->ASIL, the more QM components can interfere with the ASIL one, the lower the MTBF.
A QM baseline might not be enough to meet availability requirements.

Reviewer: Sorry but you are wrong here. A detected interference is not a failure (for fail safe systems) of the system and hence it does not contribute to the final MTBF target. For safe operational then the story is different as detecting the interference is not a good-enough mechanism (but then you need to disambiguate this clearly as usually availability is considered separate from safety  )

Reviewer: As above it all depends on your system, on your integration context. 
If for the integrator it is safety critical to avoid interference, since the only detection would still lead to a hazard, then you're right.
If instead detecting interference triggers the safe state that relies on a backup system, for example, pulling over the car, then you're not right.
So I think we should better clarify the assumptions on the integration context...

Author: The backup system is out of the scope of this document.
You can very well consider it as confirmation of my point.
This is talking about availability of the intended functionality of the "product", but let's agree it to be a car.
A pull over is a failure. It's not a safety failure, certainly. But it IS a failure of the product.
We can add a clarification about what MTBF means in this context (what I just wrote), if it can help.

## **Considerations - Toward the safety goals**
Considerations based on the two previous sections.

1. Devise negative testing simulating interference toward safety-relevant components.

> 2. Rely only on external components/mechanisms that already qualified at same-o-better safety level.

Reviewer: rather than "external" I would say "independent"

Author: can you name an independent internal component?

Reviewer: The combination of a Runtime Verification Monitor and an external WTD can be used as an independent monitor for example.

Author: anyways the RVM out of the box doesn't provide that functionality, you have to feed it the models

Reviewer: Yes of course, you need to design your specific RVM. It was just an example to justify independent as a better word compared to external....

> 3. Analyse effects of interference based on recipient (e.g. corrupted data harder to notice)

Reviewer: Here I would say "Make sure that the interference modes are avoided or detected either at the source or at the recipient side with sufficient level of rigor"

Author: No, it's intentionally excluding the source  because the whole point is that it's impossible to claim completeness about sources.

Again, if you disagree, feel free to prove me wrong and eenumerate all the possible sources.

4. When possible, leverage system configuration (e.g. external safety island / external monitor)
5. Pick adequate granularity for verification (e.g. fine grained component validation vs end-to-end)

> 6. In case of end-to-end verification, choose strategy: periodic test injection, or on demand.

Reviewer: IMO point 6 is about "Make sure that mechanisms used to prevent or detect interference are properly verified. In case of HW mechanisms periodic or on-demand diagnostic tests may be required"

Author: this is not about HW - not necessarily

it's about using QM or less SW while performing E2E verification that it's still functioning correctly

Reviewer: But then rather than e2e verification it seems an actual ASIL decomposition (and in this case we need to make sure that an appropriate safety analysis is done on the QM element to make sure the ASIL monitors cover all the dangerous failure modes)....

Author: if one end can make asusmptions about the other end, and validate them for correctness, isn't it e2e?
